webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlUserDetailsService.java

```
@@ -47,7 +47,15 @@ public Object loadUserBySAML(final SAMLCredential credential) throws UsernameNot
             final String message = String.format("Grants do not permit access to tenant for user \"%s\"", username);
             logger.warn(message);
 
-            throw new UsernameNotFoundException(message);
+            return User.builder()
+                    .id(id)
+                    .firstName(firstName)
+                    .lastName(lastName)
+                    .email(email)
+                    .username(username)
+                    .password(REDACTED)
+                    .enabled(false)
+                    .build();
```

So I'm no longer throwing the exception if a user doesn't have any grants to the tenant, and instead returning a disabled user.  

The problem with throwing the exception is that then the user never gets added to the SecurityContextHolder so then we can never log the user out correctly.  Technically the user is authenticated, just not authorized here.

---

This is kind of an abuse of the purpose of the enabled flag. It is technically for SSO systems where an admin can go in and turn a user account off and on. Similar to "locked" but not related to password retry failure.
I still think we should throw an exception here and I think we can configure an auth failure handler in the SamlWebSecurityConfiguration to handle the view routing logic

---

I'm all for it if you can get that to work.  I have been unsuccessful with it, and frankly, I've struggled to find examples online where others throw exceptions in this service and how they are handled.

I can also create a new property on the `User` and use that instead.  Something like `hasAccessToTennant` ?

---

I think ```getPermissionsById()``` will be empty if they don't have tenant access so we could check for that too. Let me fiddle with the failure handler and let you know what I find.